File: MiddleOfTheLinkedList.py
# Runtime: 49ms - Beats 64.78%
# Memory: 13.9MB - Beats 55.85%
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        mid: Optional[ListNode] = head

        while head and head.next:
            head = head.next.next
            mid = mid.next
        
        return mid

File: MiddleOfTheLinkedList.cpp
// Runtime: 0ms - Beats 100%
// Memory: 7MB - Beats 73.38%
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* mid = head;
        
        while (head && head->next){
            head = head->next->next;
            mid = mid->next;
        }
        return mid;
    }
};
